### PR TITLE
[TOPI] Access topi::matmul from Python

### PR DESCRIPTION
**Pull request by grwlf**
By this PR we move `topi::matmul` from `nn.h` to `transform.h` and add a Python-wrapper for it.

### CODE REVIEW

**Comment by tqchen** [comment]
please request reviews from the reviewers

**Comment by grwlf** [comment]
@kazum @merrymercy Could you please look into this PR? Am I understand it correctly that we are in the process of rewriting topi from Python to C++ with thin Python API?

File: topi/src/topi.cc [hunk 1]
```diff
@@ -206,6 +206,11 @@ TVM_REGISTER_GLOBAL("topi.nn.pad")
   *rv = pad(args[0], args[1], args[2], args[3]);
   });
 
+TVM_REGISTER_GLOBAL("topi.nn.matmul")
```
Comment (merrymercy) [inline topi/src/topi.cc]: matmul is not nn specific. It should be `topi.matmul`

File: topi/tests/python_cpp/test_topi_matmul.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    s = tvm.create_schedule([out.op])
+    m = tvm.build(s, pls + [out], "llvm")
+    m(*(vals_nd+[out_nd]))
+    return out_nd.asnumpy()
+
+def verify_matmul(sa, sb):
+    a = np.random.uniform(low=-1.0, high=1.0, size=get_const_tuple(sa)).astype(np.float32)
+    b = np.random.uniform(low=-1.0, high=1.0, size=get_const_tuple(sb)).astype(np.float32)
+    c1 = np.matmul(a, b)
+    c2 = with_tvm(lambda A,B: topi.cpp.nn.matmul(A,B), a,b)
```
Comment (merrymercy) [inline topi/tests/python_cpp/test_topi_matmul.py]: Add a python wrapper so we can call `topi.matmul` directly without `.cpp`

File: topi/tests/python_cpp/test_topi_matmul.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    out = lam(*pls)
+    out_nd = tvm.nd.array(np.zeros(get_const_tuple(out.shape), dtype=out.dtype), ctx)
+    s = tvm.create_schedule([out.op])
+    m = tvm.build(s, pls + [out], "llvm")
+    m(*(vals_nd+[out_nd]))
+    return out_nd.asnumpy()
+
+def verify_matmul(sa, sb):
+    a = np.random.uniform(low=-1.0, high=1.0, size=get_const_tuple(sa)).astype(np.float32)
```
Comment (merrymercy) [inline topi/tests/python_cpp/test_topi_matmul.py]: `get_const_tuple` is redundant here.

File: topi/src/topi.cc (hunk 1 above)
Comment (grwlf) [reply topi/src/topi.cc]: Actually, in C++ matmul defined in nn.h which is strange. I assume I should move it too. Could you suggest new location for matmul? How about `transform.h` / `transform.py` ?

File: topi/tests/python_cpp/test_topi_matmul.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import os
```
Comment (kazum) [inline topi/tests/python_cpp/test_topi_matmul.py]: No need to import os.

File: topi/tests/python_cpp/test_topi_matmul.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import numpy as np
+import tvm
+import topi
+from topi.util import get_const_tuple
+
+def with_tvm(lam, *args):
+    ctx = tvm.cpu(0)
+    pls = []     # placeholders
+    vals_nd = [] # initial values
+    for i in range(len(args)):
```
Comment (kazum) [inline topi/tests/python_cpp/test_topi_matmul.py]: Using enumerate looks cleaner.

File: topi/tests/python_cpp/test_topi_matmul.py (hunk 4 above)
Comment (grwlf) [reply topi/tests/python_cpp/test_topi_matmul.py]: Done

File: topi/tests/python_cpp/test_topi_matmul.py (hunk 5 above)
Comment (grwlf) [reply topi/tests/python_cpp/test_topi_matmul.py]: Done

File: topi/tests/python_cpp/test_topi_matmul.py (hunk 3 above)
Comment (grwlf) [reply topi/tests/python_cpp/test_topi_matmul.py]: Done

File: topi/tests/python_cpp/test_topi_matmul.py (hunk 2 above)
Comment (grwlf) [reply topi/tests/python_cpp/test_topi_matmul.py]: Done

File: topi/src/topi.cc (hunk 1 above)
Comment (grwlf) [reply topi/src/topi.cc]: I've moved it to transform.h for now, please let me know if it is not OK.

File: topi/python/topi/transform.py [hunk 6]
```diff
@@ -338,3 +338,21 @@ def take(a, indices, axis=None):
     if axis is None:
         return cpp.take(a, indices)
     return cpp.take(a, indices, int(axis))
+
+@tvm.tag_scope(tag=tag.INJECTIVE)
```
Comment (merrymercy) [inline topi/python/topi/transform.py]: The fuse pattern is OUT_ELEMWISE_FUSABLE
https://github.com/dmlc/tvm/blob/edd378007ddb1adb5451f2a73f60f6ff538b37a8/nnvm/python/nnvm/top/nn.py#L77

Comment (grwlf) [reply topi/python/topi/transform.py]: Very interesting place, I will try to figure it out. For now I've fixed the annotation as you suggested, Thanks!

Comment (grwlf) [reply topi/python/topi/transform.py]: I feel I really need to understand how does patterns/tags/schedulers work in order to do the right thing. Am I correct saying that `@tvm.tag_scope` does nothing when used with `cpp` functions? I see  `tvm/api.py` alone refers to `TagScope` , but in our case the C++ version of `tvm::compute` is used.

Comment (grwlf) [reply topi/python/topi/transform.py]: AFAIK, in cpp case, the default `tag` arguments are used for tagging, e.g. https://github.com/dmlc/tvm/pull/1744/commits/9784c9e06dfc383cb6e6ca70100c38aac24a0252#diff-6bbbb63b8fa5dc215443d88cfafed27aR650 If I am correct, then we probably should remove all `@tvm.tag_scope` declarations from cpp wrappers declared in `topi.transform.py` .

Comment (merrymercy) [reply topi/python/topi/transform.py]: You are right. We should remove all tag scopes for cpp wrappers.

**Comment by tqchen** [comment]
Thanks @kazum @merrymercy @grwlf , this is now merged

**Pull request merged**